tests/tests.fish

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env fish
+
+set BASE_DIR (dirname (status -f))
+
+function test_program
+    set TEST_SCRIPT $argv[1]
+    set TEST_MODE $argv[2]
+    set TEST_ARG $argv[3]
+    set TEST_EXPECTED $argv[4]
+
+    set tmpfile (mktemp)
+    echo "$TEST_SCRIPT" >"$tmpfile"
+
+    # Fish shell's '$_' variable doesn't behave the same when sourcing from a command 
+    # substitution, so setting an environment variable is a hacky workaround
+    if test "$TEST_MODE" = "source"
+        set -x IS_SOURCED true
+        set TEST_OUT (fish -c "source $tmpfile $TEST_ARG" | string collect)
+        set -e IS_SOURCED
+    else if test "$TEST_MODE" = "direct"
+        chmod +x "$tmpfile"
+        set TEST_OUT (fish -c "$tmpfile $TEST_ARG" | string collect)
+    else if test "$TEST_MODE" = "envgrep"
+        set -x IS_SOURCED true
+        set TEST_OUT (fish -c "source $tmpfile $TEST_ARG; env | grep TEST_AWS_CLI_MFA" | string collect)
+        set -e IS_SOURCED
+    else
+        echo "BAD TEST ARG!"
+        return
+    end
+
+    rm "$tmpfile"
+
+    if test "$TEST_OUT" = "$TEST_EXPECTED"
+        echo "SUCCESS - $TEST_MODE: $TEST_ARG"
+    else
+        echo "$TEST_EXPECTED"
+        echo "FAILURE - $TEST_MODE: $TEST_ARG: $TEST_OUT"
```

Comment:
It might seem unnecessary in context, but especially for consistency, as well as for potential future proofing, I would actually prefer that you retain the `TEST_SHELL` variable, just setting it to "fish" ofc. And I mention this here because these lines are missing that these are fish tests -- again, clear when you are deliberately running this script, but if this was a part of CI or whatever, it might not be as obvious at some point.

Reply:
Fixed this.